sklearn/neighbors/base.py

```diff
@@ -347,28 +349,24 @@ class from an array representing our data set and ask who's
 
         n_jobs = _get_n_jobs(self.n_jobs)
         if self._fit_method == 'brute':
+
+            reduce_func = partial(self._reduce_func, n_neighbors=n_neighbors,
+                                  return_distance=return_distance)
+
             # for efficiency, use squared euclidean distances
             if self.effective_metric_ == 'euclidean':
-                dist = pairwise_distances(X, self._fit_X, 'euclidean',
-                                          n_jobs=n_jobs, squared=True)
+                result = pairwise_distances_reduce(
+                    X, self._fit_X, 'euclidean', n_jobs=n_jobs,
+                    reduce_func=reduce_func, block_size=1, squared=True)
             else:
-                dist = pairwise_distances(
+                result = pairwise_distances_reduce(
                     X, self._fit_X, self.effective_metric_, n_jobs=n_jobs,
+                    reduce_func=reduce_func, block_size=1,
                     **self.effective_metric_params_)
-
-            neigh_ind = argpartition(dist, n_neighbors - 1, axis=1)
-            neigh_ind = neigh_ind[:, :n_neighbors]
-            # argpartition doesn't guarantee sorted order, so we sort again
-            neigh_ind = neigh_ind[
-                sample_range, np.argsort(dist[sample_range, neigh_ind])]
-
+            result = np.hstack(list(result))
```

Comment:
I think this merges the distances and neighbors into one array, and upcasts dtypes if necessary. I can't say I much like that.